### DISCUSSION TITLE
Building a FreeBSD port for owncast

### DISCUSSION THREAD

**Discussion by ctsde-markus** [owncast/owncast#944] [Ideas] [answered]
Hi!

I try to do a FreeBSD port for owncast and i may have some dumb questions that i didn't find answered anywhere (pointers are OK too, i'm lame at finding things on the internet).

I downloaded the 0.0.6 release and did a `go build. it built just fine and i can run the resulting owncast binary (and it even works, yay!).

The goal is:

* create a user owncast:owncast with it's homedir most probably in `/usr/local/www/owncast`
* copy the resulting owncast binary to /usr/local/bin (the traditional path for port binaries in FreeBSD)
* create a rc script to autostart owncast at boottime (i'll adapt that from another go port)

My first questions are:

* What needs to be in `/usr/local/www/owncast` so it runs without problems?
  * `backup/`
  * `config.yaml(.sample)`
  * `data/`
  * anything else?
* How do i tell owncast that all it needs (besides it's own binary) is in there?

**Comment by gabek** [comment]
Basically everything that is in a release version will be expected to be in one place.

- webroot
- static
- data (will be automatically created if it doesn't exist)
 
There are no config files anymore, so you don't have to worry about those.

**Reply by ctsde-markus** [reply]
OK thanks!

I'll play around with it some more. So far i just copied over the whole un-tar-ed release archive. I already created the rc script, but it's still rough (some static paths) and not ready yet for releasing it to the world, but i'm making progress. It's still a long way to go until i can create a FreeBSD port out of what i have so far.

_Do i need to copy over the admin directory too with the executable?_
Edit: There is no admin directory, ignore this question...

**Comment by ctsde-markus** [comment]
Currently i narrowed it down to a (rough) update script. I'll improve it after the next owncast release and rebuild it, so it also works as an install script then.

What i have so far:

```sh
# change VERSION to desired release
# look at https://github.com/owncast/owncast/releases
export VERSION="0.0.6"
export OWNCAST_PATH="/usr/local/www/owncast"
service owncast stop
mkdir -p /tmp/owncast
cd /tmp/owncast
fetch -o "owncast-${VERSION}.tar.gz" "https://github.com/owncast/owncast/archive/refs/tags/v${VERSION}.tar.gz"
tar -xzpvf "/tmp/owncast/owncast-${VERSION}.tar.gz"
cd "owncast-${VERSION}"
go build
chown -R owncast:owncast .
chmod 0750 .
chmod 0750 ./owncast
rm -rf ${OWNCAST_PATH}/owncast ${OWNCAST_PATH}/static ${OWNCAST_PATH}/webroot
cp -rp static webroot ${OWNCAST_PATH}/
cp -p owncast config-default.yaml LICENSE README.md CODE_OF_CONDUCT.md openapi.yaml ${OWNCAST_PATH}/
service owncast start
```

rc script `/usr/local/etc/rc.d/owncast` for easy starting it as a daemon

```sh
#!/bin/sh

# PROVIDE: owncast
# REQUIRE: NETWORKING SYSLOG
# KEYWORD: shutdown
#
# Add the following lines to /etc/rc.conf to enable owncast:
#
#owncast_enable="YES"

. /etc/rc.subr

name="owncast"
rcvar="owncast_enable"

load_rc_config $name

: ${owncast_user:="owncast"}
: ${owncast_enable:="NO"}
: ${owncast_path:="/usr/local/www/owncast"}
: ${owncast_facility:="daemon"}
: ${owncast_priority:="debug"}

command="${owncast_path}/${name}"
procname="${owncast_path}/${name}"

pidfile="/var/run/${name}.pid"

start_cmd="${name}_start"

owncast_start() {
        cd ${owncast_path}
        /usr/sbin/daemon -S -l ${owncast_facility} -s ${owncast_priority} -T ${name} \
                -u ${owncast_user} -p ${pidfile} \
                /usr/bin/env -i \
                "HOME=${owncast_path}" \
                "PATH=/usr/local/bin:${PATH}" \
                "USER=${owncast_user}" \
                $command
}

run_rc_command "$1"
```

**Reply by gabek** [reply]
This is awesome!  As an aside, would it be helpful for me to start building and releasing FreeBSD binary builds?  I'm happy to add it to our release build list of supported platforms.  Or is building it from source fine/better for your needs?

**Reply by ctsde-markus** [reply]
Nah! For FreeBSD its really better to integrate it in the ports collection directly.

It'll take some time for me, as i only played around here and there a bit with some ports. I have to find someone from the FreeBSD folks who can mentor me bit to create a port for it. But i doubt that a FreeBSD binary build would help that much.

Then it's just a matter of

```
pkg install owncast
sysrc owncast_enable="YES"
service owncast start
```

to have it up and running. That's my goal.

Even if someone is building his own packages via `poudriere` or do a manual build on his own FreeBSD server with `cd /usr/ports/www/owncast && make install` - it would all result in a working owncast server.

FreeBSD is similar to Linux for sure, but it's very different at the same time ;-)

**Reply by ctsde-markus** [reply]
I have a somehow working but still not really correct FreeBSD port ready. Stay tuned!

**Comment by ctsde-markus** [comment]
Woohooo!

I have (at least) a "works for me" FreeBSD port of owncast. It correctly builds a package on my poudriere package server for FreeBSD 11 and 12 (and it also installes correct and even runs!).

Tomorrow i'll create the patch to propose it to the FreeBSD community and then we have to wait for someone to have a look over it and eventually commit it. I have no idea how long this will take, i never proposed a new port so far.

I'll post the link here after i proposed the patch on FreeBSD's Bugzilla.

*fingers crossed*

**Reply by gabek** [reply]
So exciting!  Thanks for doing this, that's super cool that it'll be nice and easy for FreeBSD users to get it installed via their native ports.

**Comment by ctsde-markus** [comment, marked as answer]
And this is the issue on FreeBSD's Bugzilla for the new www/owncast port: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=255439

**Reply by ctsde-markus** [reply]
DAMN, that was fast!

It's commited!

https://cgit.freebsd.org/ports/commit/?id=0c973ae855ea63bcfd0d17dbe0aa5475fecf178a

Now there is an official port of owncast in the FreeBSD ports collection. Hooray! 🥳

**Reply by gabek** [reply]
Congratulations on the merge!  That's so cool they got your change in so quickly, I'm impressed!

This is the officially the first OS-specific packaging of Owncast now, awesome!
